app/Http/Controllers/SearchController.php

```diff
@@ -100,15 +107,23 @@ function (Builder $query) use ($id, $escaped_term) {
 		if ($photos != null) {
 			$i = 0;
 			foreach ($photos as $photo) {
-				$return['photos'][$i] = $photo->prepareData();
-				++$i;
+				if (Session::has('UserID') || $photo->get_public()) {
+					$return['photos'][$i] = $photo->prepareData();
+					++$i;
+				}
 			}
 		}
 
 		/**
 		 * Albums
 		 */
-		$query = Album::where('owner_id', '=', $id);
+		$query = Album::where('id', '>', 0);
+		$query->where(function (Builder $query) use ($id) {
+			$query->where('owner_id', '=', $id);
+			if (Configs::get_value('public_search', '0') == '1') {
+				$query = $query->orWhere('public', '=', 1);
+			}		
+		});		
```

Comment:
Again, this is not correct:

1. It will leak hidden albums (and their contents).

2. It will leak contents from password-protected albums.

3. It will not return shared albums (that issue predates your PR).

I've been thinking as well about how to make Search safe in public mode, though I never got around to doing anything about it. But here's how I think this needs to be done:

1. The code needs to recursively traverse the albums tree from the root down, checking the read access permissions at every level. Shared albums should be included.

2. In the process, an array of `albumIDs` of accessible albums needs to be constructed. This should include all subalbums. Hidden albums or their subalbums should not be included (even if public and not hidden). The contents of password-protected albums should not be included (again, even if public and not password-protected), unless the password has already been provided.

3. Subsequent queries, both on the Albums and Photos (see above), need to use that array as a filter.

We already had such a filter: by `owner_id`, but it was not sufficient and obviously won't work in the public mode.

Actually, we already have the code that does all of the above in other places, specifically:

- `App\Http\Controllers\AlbumsController::get` extracts all the accessible top-level albums
- `App\ModelFunctions\AlbumFunctions::prepare_albums` extracts the accessible sub-albums (we only need the access check and `get_albums` from this one)

I think the right way to do it would be to refactor `App\Http\Controllers\AlbumsController::get`, moving the bulk of its code to a new method that could also be used from search.

Reply:
One more thing:
- `AlbumFunctions::get_sub_albums` can probably be useful here as, unlike `get_albums`, it returns results in the format we need for SQL queries...

Reply:
OK, thanks. Here it gets pretty obvious that I neither know Laravel generally nor Lychee particularly in depth. I tried to implement the public search with minimal invasion into the existing code - which clearly introduces some security risks. I will try to comprehend your comment and see if I can do some adjustments. 

By now, I did not realize that there now is an album hirarchy or a possibility to share between registered users. I apologize for the shortcuts.

Reply:
hahaha no worries. Indeed the v4 has multi user support (locked or not, with upload right or not) and sharing between users also available. And we also have some other plans in mind e.g. rss feed...

Reply:
I will try to improve your Laravel code. :)